### ISSUE TITLE
Committing to Branch Fails

### ISSUE THREAD

**Issue by gfs** [FuLagann/csharp-docs-generator#20]
The final step to commit the documentation seems to fail.

https://github.com/microsoft/OAT/pull/13

It appears that you check out into a detached head (https://www.git-tower.com/learn/git/faq/detached-head-when-checkout-commit) state.  This is also likely why #16 occurs.

In a detached head state pull doesn't make sense - there's nothing to pull from.

It looks also like the directory containing docs isn't being added to git.

You then try to push, which also doesn't work because you can't push to a detached head, you need to associate it with a branch.

For example, if the user specifies a branch-name you could at the start

`git checkout -b branch-name`

To check out and be on a new branch.

You'd have to decide what to do if there is already a branch with that name on origin.

```
From https://github.com/microsoft/OAT
 * [new branch]      gfs-patch-1           -> origin/gfs-patch-1
 * [new branch]      gfs/Captures          -> origin/gfs/Captures
 * [new branch]      gfs/PortFromAsa       -> origin/gfs/PortFromAsa
 * [new branch]      gfs/Sharpen           -> origin/gfs/Sharpen
 * [new branch]      gfs/VehicleDemoExpand -> origin/gfs/VehicleDemoExpand
 * [new branch]      main                  -> origin/main
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

"C:\Program Files\Git\bin\git.exe" status
HEAD detached at pull/13/merge
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	modified:   version.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	docs/api/

"C:\Program Files\Git\bin\git.exe" pull
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

"C:\Program Files\Git\bin\git.exe" pull
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

Error: The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 1
    at ExecState._setResult (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:561:25)
    at ExecState.CheckComplete (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:544:18)
    at ChildProcess.<anonymous> (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:444:27)
##[error]The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 1
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
"C:\Program Files\Git\bin\git.exe" push
fatal: You are not currently on a branch.
To push the history leading to the current (detached HEAD)
state now, use

    git push origin HEAD:<name-of-remote-branch>

Error: The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 128
    at ExecState._setResult (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:561:25)
##[error]The process 'C:\Program Files\Git\bin\git.exe' failed with exit code 128
    at ExecState.CheckComplete (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:544:18)
    at ChildProcess.<anonymous> (d:\a\_actions\FuLagann\csharp-docs-generator\v1.0\node_modules\@actions\exec\lib\toolrunner.js:444:27)
    at ChildProcess.emit (events.js:210:5)
    at maybeClose (internal/child_process.js:1021:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
```

**Comment by FuLagann** [comment]
Thank you for reporting this. I'm investigating a fix right now.

Seems like the repository gets put in a detached head when making a pull request; which is something I didn't account for. The current implementation tries to fix a missed pull which is only 1 edge case (which happens when crashing after the git push). I'll draft something up for a better system to interface with git and come up with a fix soon.

**Comment by gfs** [comment]
You may want skip doing the git magic yourself and instead recommend users use another library that handles that like: https://github.com/peaceiris/actions-gh-pages

I found this very easy to use.

Your action could then just turn code into a doc at a location, and you use the peaceiris action to publish that location.

**Comment by FuLagann** [comment]
Alright, this has been a while since I've last worked on this; sorry for such a delay.

I tested to the best of my knowledge and the [latest PR](https://github.com/FuLagann/csharp-docs-generator/pull/23) should fix it. When pull requesting, you'll need to set the `branch-name` to the same branch where the it's pulling from.

To answer your query, I had to make my own process for pushing to git or else nothing will appear in the repo. Didn't know if there was any actions related to pushing to git. There is a new input `skip-git` being added to voluntarily skip the default git process in place for your own.
